### DISCUSSION TITLE
How to hide "Proxy error: Could not proxy request ..." from the terminal

### DISCUSSION THREAD

**Discussion by stas-kh** [chimurai/http-proxy-middleware#606] [Q&A]
I am using `http-proxy-middleware` in my VueJS application as a proxy tool.

Here is my simple configuration of `devServer` in `vue.config.js`:
```
  devServer: {
    host: "localhost",
    port: 8081,
    proxy: {
      "^/api": {
        target: "http://localhost:11339",
        ws: true,
        changeOrigin: true,
      },
    },
  },
```

Recently I've introduced cancelling of some HTTP requests and it causes this kind of error in terminal when devServer is running:
```
Proxy error: Could not proxy request %request_url% from localhost:8081 to http://localhost:11339.
See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (ECONNRESET).
```

I would like to hide such errors from the terminal because I cancel requests manually and it's expected.

I've tried a couple of solutions:

1. Setting `logLevel: silent` to proxy config
2. Applying this hook to catch an error:
```
onProxyReq: (proxyReq, req, res) => {
  req.on("error", function(err) {
    console.warn("ERROR, ", err);
  });
},
```

None of the options could help me to hide these `ECONNRESET` errors.

I would appreciate if someone can help me in this question.

**Comment by chimurai** [comment]
http-proxy-middleware log statements are prefixed with `[HPM]`.

The message you're seeing is probably coming from http-proxy.

Found this undocumented `econnreset` proxy event in `@types/http-proxy`.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0ebb554617399eadb906c4262823bf9f3fcbfa6f/types/http-proxy/index.d.ts#L151-L159

This `econnreset` proxy event is not documented on:
https://github.com/http-party/node-http-proxy#listening-for-proxy-events

Can you try to patch and manually subscribe to `econnreset`?
https://github.com/chimurai/http-proxy-middleware/blob/e2043455090493d3fface37ba4d69a8d11b15553/src/http-proxy-middleware.ts#L34

```js
this.proxy.on('error', this.logError); 
this.proxy.on('econnreset', () => { // do nothing })
```

If this works I can expose the missing events.

**Comment by stas-kh** [comment]
@chimurai thanks for your quick response.

I've patched the library with the following line but it didn't help a lot:
```
proxy.on('econnreset', () => {})
```


I was able to extend my `vue.config.js` with this configuration:
```
onProxyReq: (proxyReq, req, res) => {
  req.on("error", function(err) {
    console.warn(`Error -`, err);
  });
}
```

and I see that the error gets caught:
```
Proxy error: Could not proxy request %request_url% from localhost:8081 to http://localhost:11339.
See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (ECONNRESET).

ERROR - Error: aborted
    at connResetException (node:internal/errors:621:14)
    at abortIncoming (node:_http_server:579:25)
    at socketOnEnd (node:_http_server:596:5)
    at Socket.emit (node:events:339:22)
    at endReadableNT (node:internal/streams/readable:1289:12)
    at processTicksAndRejections (node:internal/process/task_queues:80:21) {
  code: 'ECONNRESET'
}

Proxy error: Could not proxy request %request_url% from localhost:8081 to http://localhost:11339.
See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (ECONNRESET).
```

But the error messages are still displayed in the terminal.

The only one solution that helped was commenting out this line in `http-proxy` repository:
https://github.com/http-party/node-http-proxy/blob/9b96cd725127a024dabebec6c7ea8c807272223d/lib/http-proxy/passes/web-incoming.js#L165

@chimurai what would you suggest me to do about it?

Thanks in advance!

**Reply by chimurai** [reply]
The error message does look like this message, but not exactly the same.

https://github.com/chimurai/http-proxy-middleware/blob/e2043455090493d3fface37ba4d69a8d11b15553/src/http-proxy-middleware.ts#L191-L192


Which version of http-proxy-middleware are you using?

**Reply by stas-kh** [reply]
Here are the versions used:

```
http-proxy-middleware@0.19.1
http-proxy@^1.17.0
```

The text of the error is taken from `vue-cli-service` from the function `onProxyError`

**Reply by chimurai** [reply]
Are you able to replicate the error with a **minimal** setup and **latest** versions?

`express` + `http-proxy-middleware`

```js
express().use(createProxyMiddleware({
  target: "http://localhost:11339",
  ws: true,
  changeOrigin: true,
}))
.listen(8081)
```

Without an isolated example there is little to investigate.
